plugins/poll/lib/poll.rb

```diff
@@ -31,43 +41,100 @@ def self.vote(user, post_id, poll_name, options)
             end
 
-        # remove non-selected votes
-        PollVote.where(poll: poll, user: user).where.not(poll_option_id: new_option_ids).delete_all
+        if poll.type == "irv"
+          # for IRV, we need to remove all votes and re-create them as there is no way to update them due to lack of primary key.
+          PollVote.where(poll: poll, user: user).delete_all
+          creation_set = new_option_ids
+        else
+          # remove non-selected votes
+          PollVote
+            .where(poll: poll, user: user)
+            .where.not(poll_option_id: new_option_ids)
+            .delete_all
+          creation_set = new_option_ids - old_option_ids
+        end
 
         # create missing votes
-        (new_option_ids - old_option_ids).each do |option_id|
-          PollVote.create!(poll: poll, user: user, poll_option_id: option_id)
+        creation_set.each do |option_id|
+          if poll.type == "irv"
+            PollVote.create!(
+              poll: poll,
+              user: user,
+              poll_option_id: option_id,
+              rank:
+                options.find { |o| o[:digest] == poll.poll_options.find(option_id).digest }[:rank],
+            )
+          else
+            PollVote.create!(poll: poll, user: user, poll_option_id: option_id)
+          end
         end
       end
 
-    # Ensure consistency here as we do not have a unique index to limit the
-    # number of votes per the poll's configuration.
-    is_multiple = serialized_poll[:type] == "multiple"
-    offset = is_multiple ? (serialized_poll[:max] || serialized_poll[:options].length) : 1
-
-    DB.query(<<~SQL, poll_id: poll_id, user_id: user.id, offset: offset)
-    DELETE FROM poll_votes
-    USING (
-      SELECT
-        poll_id,
-        user_id
-      FROM poll_votes
-      WHERE poll_id = :poll_id
-      AND user_id = :user_id
-      ORDER BY created_at DESC
-      OFFSET :offset
-    ) to_delete_poll_votes
-    WHERE poll_votes.poll_id = to_delete_poll_votes.poll_id
-    AND poll_votes.user_id = to_delete_poll_votes.user_id
-    SQL
+    unless serialized_poll[:type] == "irv"
+      # Ensure consistency here as we do not have a unique index to limit the
+      # number of votes per the poll's configuration.
+      is_multiple = serialized_poll[:type] == "multiple"
+      offset = is_multiple ? (serialized_poll[:max] || serialized_poll[:options].length) : 1
+
+      DB.query(<<~SQL, poll_id: poll_id, user_id: user.id, offset: offset)
+      DELETE FROM poll_votes
+      USING (
+        SELECT
+          poll_id,
+          user_id
+        FROM poll_votes
+        WHERE poll_id = :poll_id
+        AND user_id = :user_id
+        ORDER BY created_at DESC
+        OFFSET :offset
+      ) to_delete_poll_votes
+      WHERE poll_votes.poll_id = to_delete_poll_votes.poll_id
+      AND poll_votes.user_id = to_delete_poll_votes.user_id
+      SQL
+    end
+
+    if serialized_poll[:type] == "irv"
```

Comment:
We should define constants for these in Poll and for multiple too

Reply:
Done